src/live_restore/live_restore_fs.c

```diff
@@ -246,18 +248,94 @@ __live_restore_fs_find_layer(WT_FILE_SYSTEM *fs, WT_SESSION_IMPL *session, const
     return (0);
 }
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
 /*
- * __live_restore_fs_notsup --
- *     Return an error message indicating the given functionality is not supported.
+ * __live_restore_fs_directory_list_worker --
+ *     The list is a combination of files from the destination and source directories. For
+ *     destination files, exclude any files matching the marker paths. For source files, exclude
+ *     files that are either marked as tombstones or already present in the destination directory.
  */
 static int
-__live_restore_fs_notsup(WT_SESSION *wt_session)
+__live_restore_fs_directory_list_worker(WT_FILE_SYSTEM *fs, WT_SESSION *wt_session,
+  const char *directory, const char *prefix, char ***dirlistp, uint32_t *countp, bool single)
 {
-    WT_RET_MSG((WT_SESSION_IMPL *)wt_session, ENOTSUP, "Unsupported fs operation");
+    WT_DECL_RET;
+    WT_LIVE_RESTORE_FS *lr_fs;
+    WT_SESSION_IMPL *session;
+    size_t dirallocsz;
+    uint32_t count_dest, count_src;
+    char **dirlist_dest, **dirlist_src, **entries, **namep, *path_dest, *path_src;
+    bool dest_exist, have_tombstone;
+
+    session = (WT_SESSION_IMPL *)wt_session;
+    lr_fs = (WT_LIVE_RESTORE_FS *)fs;
+    count_dest = count_src = dirallocsz = 0;
+    *dirlistp = dirlist_dest = dirlist_src = entries = NULL;
+    path_dest = path_src = NULL;
+    dest_exist = have_tombstone = false;
+
+    __wt_verbose_debug1(session, WT_VERB_FILEOPS, "DIRECTORY LIST %s (single ? %s) : ", directory,
+      single ? "YES" : "NO");
+
+    /* Get files from destination. */
+    WT_ERR(__live_restore_fs_backing_filename(
+      &lr_fs->destination, session, lr_fs->destination.home, directory, &path_dest));
+    WT_ERR(lr_fs->os_file_system->fs_directory_list(
+      lr_fs->os_file_system, wt_session, path_dest, prefix, &dirlist_dest, countp));
+
+    for (namep = dirlist_dest; *namep != NULL; namep++)
+        if (namep != NULL && *namep != NULL &&
+          !(strlen(*namep) >= strlen(".deleted") &&
+            strcmp(*namep + strlen(*namep) - strlen(".deleted"), ".deleted") == 0)) {
+            WT_ERR(__wt_realloc_def(session, &dirallocsz, count_dest + 1, &entries));
+            WT_ERR(__wt_strdup(session, *namep, &entries[count_dest]));
+            ++count_dest;
+
+            if (single)
+                goto done;
+        }
+
+    /* Get files from source. */
+    WT_ERR(__live_restore_fs_backing_filename(
+      &lr_fs->source, session, lr_fs->destination.home, directory, &path_src));
+    WT_ERR(lr_fs->os_file_system->fs_directory_list(
+      lr_fs->os_file_system, wt_session, path_src, prefix, &dirlist_src, countp));
+
+    for (namep = dirlist_src; *namep != NULL; namep++) {
```

Comment:
What is the value of `namep` in this loop? I have a suspicion that it'll return either `WT_SOURCE/foo.wt` or `foo.wt`, and that needs to be converted to `WT_DEST/foo.wt` before we call has_tombstone.